Framework/DataHandling/src/LoadMuonNexusV2Helper.cpp

```diff
@@ -64,18 +70,22 @@ NXInt loadGoodFramesDataFromNexus(const NXEntry &entry,
 std::vector<detid_t>
 loadDetectorGroupingFromNexus(const NXEntry &entry,
                               const std::vector<detid_t> &detectorsLoaded,
-                              bool isFileMultiPeriod) {
-
+                              bool isFileMultiPeriod, const int periodNumber) {
+  // We cast the numLoadedDetectors to an int, which is the index type used for
+  // Nexus Data sets. As detectorsLoaded is derived from a nexus entry we
+  // can be certain we won't overflow the integer type.
+  int numLoadedDetectors = static_cast<int>(detectorsLoaded.size());
   std::vector<detid_t> grouping;
+  grouping.reserve(numLoadedDetectors);
   // Open nexus entry
   NXClass detectorGroup = entry.openNXGroup(NeXusEntry::DETECTOR);
   if (detectorGroup.containsDataSet(NeXusEntry::GROUPING)) {
     NXInt groupingData = detectorGroup.openNXInt(NeXusEntry::GROUPING);
     groupingData.load();
-    if (!isFileMultiPeriod) {
-      for (const auto &detectorNumber : detectorsLoaded) {
-        grouping.emplace_back(groupingData[detectorNumber - 1]);
-      }
+    int groupingOffset =
+        !isFileMultiPeriod ? 0 : (numLoadedDetectors) * (periodNumber - 1);
+    for (const auto &detectorNumber : detectorsLoaded) {
```

Comment:
Just turn this into
`for (auto &detectorNumber:`

Since were working with primatives (or close to) so the optimiser / compiler will do the rest